Assets/BossRoom/Scripts/Client/UI/Lobby/JoinLobbyUI.cs

```diff
@@ -0,0 +1,160 @@
+        LocalLobby.LobbyData m_LocalLobbySelected;
+
+
+        [Inject]
+        private void InjectDependencies(UIFactory uiFactory, LobbyUIManager lobbyUIManager, LobbyServiceData lobbyServiceData)
+        {
+            m_UIFactory = uiFactory;
+            m_LobbyUIManager = lobbyUIManager;
+            m_LobbyServiceData = lobbyServiceData;
+        }
+
+        private void Awake()
+        {
+            m_LobbyPanelPrototype.gameObject.SetActive(false);
+        }
+
+        public void LobbyPanelSelected(LocalLobby lobby)
+        {
+            m_LocalLobbySelected = lobby.Data;
+        }
+
+        public void OnLobbyCodeInputFieldChanged(string newCode)
+        {
+            if (!string.IsNullOrEmpty(newCode))
+                m_LocalLobbySelected = new LocalLobby.LobbyData(newCode.ToUpper());
+        }
+
+        public void OnJoinButtonPressed()
+        {
+            m_LobbyUIManager.JoinLobbyRequest(m_LocalLobbySelected);
+            m_LocalLobbySelected = default;
+        }
+
+        public void OnRefresh()
+        {
+            m_LobbyUIManager.QueryLobbiesRequest();
+        }
+
+        protected override void UpdateObserver(LobbyServiceData observed)
+        {
+            ///Check for new entries, We take CurrentLobbies as the source of truth
+            var previousKeys = new List<string>(m_LobbyButtons.Keys);
+
+            foreach (var codeLobby in observed.CurrentLobbies)
+            {
+                var lobbyCodeKey = codeLobby.Key;
+                var lobbyData = codeLobby.Value;
+
+                if (!m_LobbyButtons.ContainsKey(lobbyCodeKey))
+                {
+                    if (CanDisplay(lobbyData))
+                    {
+                        CreateLobbyPanel(lobbyCodeKey, lobbyData);
+                    }
+                }
+                else
+                {
+                    if (CanDisplay(lobbyData))
+                    {
+                        UpdateLobbyButton(lobbyCodeKey, lobbyData);
+                    }
+                    else
+                    {
+                        RemoveLobbyButton(lobbyData);
+                    }
+                }
+
+                previousKeys.Remove(lobbyCodeKey);
+            }
+
+            foreach (var key in previousKeys)
+            {
+                // Need to remove any lobbies from the list that no longer exist.
+                RemoveLobbyButton(m_LocalLobby[key]);
+            }
+        }
+
+        public void JoinMenuChangedVisibility(bool show)
+        {
+            if (show)
+            {
+                m_JoinCodeField.text = "";
+                OnRefresh();
+            }
+        }
+
+        public void OnQuickJoinClicked()
+        {
+            m_LobbyUIManager.QuickJoinRequest();
+        }
+
+        private bool CanDisplay(LocalLobby lobby)
+        {
+            return lobby.Data.State == LobbyState.Lobby && !lobby.Private;
```

Comment:
Private lobbies are not returned in search results so it's not necessary to check that property.

Reply:
Another thought regardling `State`, can we consider including `State` as a filterable property in the room.  Then when you query, you can filter only to rooms in the state you care about, and you don't need to do any client-side filtering like this.  It should simplify the display logic.